src/main/java/org/wildfly/security/auth/server/RealmIdentity.java

```diff
@@ -266,4 +284,28 @@ public boolean exists() throws RealmUnavailableException {
             return false;
         }
     };
+
+    /**
+     * Represents a key that is specific to the realm and used to uniquely identify a {@link RealmIdentity} instance.
+     */
+    interface Key {
+
+        Key EMPTY = (Key) () -> null;
+
+        /**
+         * Returns the value of this key.
+         *
+         * @return the value of this key
+         */
+        Object getValue();
```

Comment:
The reason this works - when it seems like we've been wanting to get rid of Principal from the realm - is that we really would be using Principal for two things: identifying an identity within a domain, and identifying an entry within a realm.  The realm principal is not really relevant to the domain, just like the domain principal should never have been relevant to the realm.

Really I think this solution is much cleaner because we then don't have N ways to find an identity, we just have 2 ways.  With the caveat that every realm needs to support NamePrincipal (for now anyway).